pkg/operator/configobserver/node/suppress_config_updates.go

```diff
@@ -110,3 +115,112 @@ func (s *revisionDiffProfileSuppressor) shouldSuppressConfigUpdates() (suppress
+// It checks the current value of observedConfig from operator's state (static pod operator resource), tries to determine the latency profile
+// that was applied by the config observer and compares the same against the desired profile set on the cluster. In case, the update from
+// the active profile to the desired profile is an extreme scenario i.e. one of the scenarios specified in rejectionScenarios,
+// the suppressor suppresses the update.
+func NewSuppressConfigUpdateForExtremeProfilesFunc(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	configNodeLister configlistersv1.NodeLister,
+	latencyConfigs []LatencyConfigProfileTuple,
+	rejectionScenarios []LatencyProfileRejectionScenario,
+) ShouldSuppressConfigUpdatesFunc {
+
+	usedConfigPaths := make([][]string, len(latencyConfigs))
+	for i, latencyConfig := range latencyConfigs {
+		usedConfigPaths[i] = latencyConfig.ConfigPath
+	}
+
+	ret := rejectExtremeProfilesSupressor{
+		operatorClient:     operatorClient,
+		configNodeLister:   configNodeLister,
+		latencyConfigs:     latencyConfigs,
+		usedConfigPaths:    usedConfigPaths,
+		rejectionScenarios: rejectionScenarios,
+	}
+	return ret.shouldSuppressConfigUpdates
+}
+
+func (s *rejectExtremeProfilesSupressor) shouldSuppressConfigUpdates() (bool, error) {
+	// get current state of the config
+	operatorSpec, _, _, err := s.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return false, err
+	}
+
+	// raw bytes of observed config in JSON format
+	observedConfigRaw, err := kyaml.ToJSON(operatorSpec.OperatorSpec.ObservedConfig.Raw)
+	if err != nil {
+		return false, err
+	}
+
+	// decode the JSON
+	var observedConfig map[string]interface{}
+	err = json.Unmarshal(observedConfigRaw, &observedConfig)
+	if err != nil {
+		return false, err
+	}
+	// prune the obtained config with only config paths that we monitor for latency profiles
+	observedConfigPruned := configobserver.Pruned(observedConfig, s.usedConfigPaths...)
+
+	knownProfileConfigs := make(map[configv1.WorkerLatencyProfileType]map[string]interface{})
+	// traverse along each latency config to construct a config arg val map for each known profile
+	for _, latencyConfig := range s.latencyConfigs {
+		for profile := range latencyConfig.ProfileConfigValues {
+			profileValue := latencyConfig.ProfileConfigValues[profile]
+
+			_, ok := knownProfileConfigs[profile]
+			if !ok {
+				knownProfileConfigs[profile] = make(map[string]interface{})
+			}
+			err = unstructured.SetNestedStringSlice(knownProfileConfigs[profile], []string{profileValue}, latencyConfig.ConfigPath...)
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+
+	// determine which profile was in the obtained config
+	var fromProfile configv1.WorkerLatencyProfileType
+	for profile := range knownProfileConfigs {
+		if reflect.DeepEqual(observedConfigPruned, knownProfileConfigs[profile]) {
+			fromProfile = profile
+			break
+		}
+	}
```

Comment:
Yeah, added:

```
       // in case profile present in configmap couldn't be determined
	if fromProfile == "" {
		return false, "", nil
	}
```

I guess we should not use `return false, "", fmt.Errorf("profile not found")` because that could degrade our controller.

Reply:
Eg. the config map(s) being checked here are across known values do not merge with UnsupportedConfigOverrides during comparison, so in case of UnsupportedConfigOverrides, `fromProfile` will be undeterminable which would degrade the controller, if we do not pass an error here, we'll be stuck in progressing instead of degraded atleast.

IIRC, we had a similar discussion when wiring status updates for the controller itself with the `matchRevisionFn` where we also do not check for overrides while comparing config maps, and allow the controller always be in Progressing state as it is unsupported by our operator(s) :)

Reply:
If you check the updated invocation of the controller and observer in KCMO in [the corresponding PR](https://github.com/openshift/cluster-kube-controller-manager-operator/pull/629), the `LatencyProfileRejectionScenarios` are defined as [1]: 

```go
var LatencyProfileRejectionScenarios = []nodeobserver.LatencyProfileRejectionScenario{
	{FromProfile: "", ToProfile: configv1.LowUpdateSlowReaction},
	{FromProfile: configv1.LowUpdateSlowReaction, ToProfile: ""},

	{FromProfile: configv1.DefaultUpdateDefaultReaction, ToProfile: configv1.LowUpdateSlowReaction},
	{FromProfile: configv1.LowUpdateSlowReaction, ToProfile: configv1.DefaultUpdateDefaultReaction},
}
```

where the "" (empty) -> Low and Low -> "" (empty) is also considered as a rejection scenario, and essentially:
- for observers, the suppressors are run even before any actual config observation code is run (so "" is valid profile during suppression)
- for controller, the `checkRejectionFn` is also called before matching revisions or updating profile rollout status, so similarly "" (empty) profile would be handled too

[1] https://github.com/openshift/cluster-kube-controller-manager-operator/pull/629/files#diff-6470c27b84831e97d7885b3be1d1de998c7f724d8163e5952ca1f978264f9ef9R20-R26